pkg/client/client.go

```diff
@@ -104,14 +107,19 @@ func (client *DataSetClient) ListenAndSendBufferForSession(session string, ch ch
 			client.Logger.Debug("Received buffer from channel",
 				zap.String("session", session),
 				zap.Int("index", i),
+				zap.Uint64("buffersEnqueued", client.buffersEnqueued.Load()),
+				zap.Uint64("buffersProcessed", client.buffersProcessed.Load()),
 			)
 			buf, ok := msg.(*buffer.Buffer)
+			client.workers.Add(1)
 			if ok {
 				for client.RetryAfter().After(time.Now()) {
 					client.sleep(client.RetryAfter(), buf)
 				}
 				if !buf.HasEvents() {
 					client.Logger.Warn("Buffer is empty, skipping", buf.ZapStats()...)
+					client.workers.Done()
```

Comment:
Do we have test cases for all the scenarios / code paths which ensure that those counters are correctly incremented / decremented in all the code paths so we can't end up in an infinite loop where `AddEvents` function would never return?

Reply:
Function `AddEvents` always returns. This additional waiting is now in the function `Finish`. Usage could be seen in the tests.